nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,535 @@
+// Copyright (c) 2022 Samsung Research Russia
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/collision_monitor_node.hpp"
+
+#include <exception>
+#include <cmath>
+#include <limits>
+#include <utility>
+
+#include "tf2_ros/create_timer_ros.h"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+CollisionMonitorNode::CollisionMonitorNode(const rclcpp::NodeOptions & options)
+: nav2_util::LifecycleNode("collision_monitor_node", "", false, options),
+  velocity_({0.0, 0.0, 0.0}), velocity_valid_(false), cmd_vel_in_({0.0, 0.0, 0.0}),
+  ra_prev_({DO_NOTHING, {0.0, 0.0, 0.0}}), release_operation_(false)
+{
+  RCLCPP_INFO(get_logger(), "Creating CollisionMonitorNode");
+}
+
+CollisionMonitorNode::~CollisionMonitorNode()
+{
+  RCLCPP_INFO(get_logger(), "Destroying CollisionMonitorNode");
+  polygons_.clear();
+  sources_.clear();
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  // Transform buffer and listener initialization
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    this->get_node_base_interface(),
+    this->get_node_timers_interface());
+  tf_buffer_->setCreateTimerInterface(timer_interface);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_); 
+
+  // Obtaining ROS parameters
+  if (!getParameters()) {
+    return nav2_util::CallbackReturn::FAILURE;
+  }
+
+  // Setting subscribers and publishers
+  rclcpp::QoS odom_qos = rclcpp::SystemDefaultsQoS();  // set to default
+  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+    odom_topic_, odom_qos,
+    std::bind(&CollisionMonitorNode::odomCallback, this, std::placeholders::_1));
+
+  cmd_vel_in_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
+    cmd_vel_in_topic_, 1,
+    std::bind(&CollisionMonitorNode::cmdVelInCallback, this, std::placeholders::_1));
+  cmd_vel_out_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
+    cmd_vel_out_topic_, 1);
+
+  if (polygon_topic_.length()) {
```

Comment:
```suggestion
  if (polygon_topic_.length() > 0) {
```

Reply:
I know not required, but is more semantically readable

Reply:
In latest changes, do not need: we are checking `visualize` parameter instead.